src/app/blog/iroh-dag-sync/page.mdx

```diff
@@ -0,0 +1,293 @@
+two nodes. You can think of it as [sendme], but for dags.
+
+## Generate some data
+
+We need a car file. You can just import some directory into ipfs and then export
+it as a car file. Make sure to use --raw-leaves to have a more interesting dag.
+
+Let's use the linux kernel sources as an example.
+
+```
+> ipfs add ../linux --raw-leaves
+> ipfs dag export QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV > linux.car
+```
+## Import the data
+
+```
+> cargo run --release import linux.car
+...
+root: QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV
+```
+
+This will create two databases in the current directory. dag.db contains
+information about the structure of the dag, blobs.db (a directory) contains
+the raw data.
+
+- Start a node that makes the data available
+
+```
+> cargo run --release node
+I am irgkesdtbih664hq2fjgd6zf7g6mazqkr7deqzplavmwl3vdbboa
+```
+
+## Sync with the default traversal
+
+In a *different directory*, start the sync process:
+
+```
+> mkdir tmp
+> cd tmp
+> cargo run --release sync --from irgkesdtbih664hq2fjgd6zf7g6mazqkr7deqzplavmwl3vdbboa QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV
+```
+
+This will traverse the entire DAG in depth-first, pre-order, left-to-right
+traversal order. Which may take a while. But - it is just a single request/
+response pair, so we will saturate the wire.
+
+## Export the synced data (optional)
+
+```
+> cargo run --release export QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV --target output.car
+```
+Export without specifying a target just dumps the cids to stdout.
+
+## Advanced use
+
+The above example sync a large dag using a single request/response interaction.
+It uses the default strategy of traversing the dag in depth first, preorder,
+left to right, which does the job if you want to sync the entire dag.
+
+But what if you want to do a partial sync? You can specify more complex
+configurations for the existing defined traversals.
+
+### Exclude leaf nodes
+
+```
+cargo run --release sync --from bsmlrj4sodhaivs2r7tssw4zeasqqr42lk6xt4e42ikzazkp4huq \
+  --traversal 'Full(root:"QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV",filter:NoRaw)'
+```
+
+Here we specify the traversal to use as a [ron] expression, specifying a full
+traversal where we filter out cids that have format raw and are guaranteed to be
+leaf nodes of the dag.
+
+In case of unixfs data, the vast majority of the bytes of the dag are in the leaf
+nodes, so requesting only non-leaf nodes gives us the ability to inspect the
+directory structure while saving a lot of bandwidth.
+
+We could also, in a second step, request all leaves from multiple senders.
+
+### Full traversal, but exclude data for leaf nodes.
+
+```
+cargo run --release sync --from bsmlrj4sodhaivs2r7tssw4zeasqqr42lk6xt4e42ikzazkp4huq \
+  --traversal 'Full(root:"QmWyLtd4WEJe45UBqCZG94gYY9B8qF3k4DKFX3o2bodHmV")' --inline NoRaw
+```
+
+In this case we do a full traversal, but only *inline data* for non-raw blocks that
+can possibly contain links.
```

Comment:
Can you add another line or two describing the difference between this mode and the one from the previous section?

Is it that in the previous mode, we don't even send over the CIDs for the raw nodes, but in this mode we DO send over the CIDs for the raw nodes, but just don't fetch the data corresponding to those raw format CIDs?

In other words, we get all the CIDs for the DAG, but only the data for the non-raw CIDs. In the previous section we only get the CIDs for the non-raw CIDs?

Reply:
Yes, exactly. The idea is that you have the cid to blake3 mapping and can then try to get the data via blake3 hash from somewhere else (you can not trust the mapping though until you have downloaded and validated with the non blake3 hash!)